src/Access/Credentials.h

```diff
@@ -85,6 +85,42 @@ class BasicCredentials
     bool allow_interactive_basic_authentication_in_the_browser = false;
 };
 
+class BearerCredentials
+    : public Credentials
+{
+public:
+    explicit BearerCredentials(const String & user_name_)
+    {
+        user_name = user_name_;
+        is_ready = !user_name.empty();
+    }
+
+    void setToken(const String & token_) { token = token_; }
+    const String & getToken() const { return token; }
+
+private:
+    String token;
+};
+
+class JWTCredentials
+    : public Credentials
+{
+public:
+    explicit JWTCredentials() = default;
+    explicit JWTCredentials(const String & token_);
+    explicit JWTCredentials(const String & token_, const String & jwks_);
+
+    void setToken(const String & token_) { token = token_; }
+    void setJWKS(const String & jwks_) { jwks = jwks_; }
+    const String & getToken() const { return token; }
+    bool isValid() const { return valid; }
+
+private:
+    String token;
+    String jwks;
```

Comment:
Either I do not understand the logic, or it is a strange one
Are you planning to set JWKS per-user?

Reply:
This is an incorrect approach then: JWK(S) is simply an extended public key (or a set of them). They are managed by a third party service (e.g. Keycloak) or a third party app in your server, but not ClickHouse itself. They are not a per-user thing, they are per-issuer thing. Even if you want to manage JWK(S) manually, it does not make sense to store it in Credentials - it is a system-wide thing, very alien from user data

Reply:
> There could be more then one system with self assigned JWTs in organization. And we do not want to duplicate DBMS users for them

Then usually you would have not one JWK to validate against, but several. Then you can authenticate the same user with JWTs created and signed by different entities

Reply:
> There could be more then one system with self assigned JWTs in organization. And we do not want to duplicate DBMS users for them

Then usually you would have not one JWK to validate against, but several. Then you can authenticate the same user with JWTs created and signed by different entities

Reply:
You are right that it is not per user thing. I should split it in XML: one JWKS section in config for all users and 'use_jwt' in per-user section.

What about external nature of JWKS. It's the same for all hash auths. There should be some way to place external auth keys into DBMS. There's no troubles to set them in config file and pass to user auth phase any way DBMS needed.